src/main/java/com/commercetools/project/sync/SyncerFactory.java

```diff
@@ -55,122 +61,172 @@ public static SyncerFactory of(
-            })
-        .thenCompose(
-            ignored ->
-                ProductSyncer.of(sourceClient, targetClient, clock)
-                    .sync(runnerNameOptionValue, isFullSync))
-        .thenCompose(
-            ignored ->
-                ShoppingListSyncer.of(sourceClient, targetClient, clock)
-                    .sync(runnerNameOptionValue, isFullSync))
-        .whenComplete((syncResult, throwable) -> closeClients());
-  }
+    final List<SyncModuleOption> syncModuleOptions;
+    try {
+      syncModuleOptions = validateAndCollectSyncOptionValues(syncOptionValues);
+    } catch (IllegalArgumentException exception) {
+      return exceptionallyCompletedFuture(exception);
+    }
 
-  private void closeClients() {
-    sourceClientSupplier.get().close();
-    targetClientSupplier.get().close();
+    final Collection<List<SyncModuleOption>> groupedSyncModuleOptions =
+        groupSyncModuleOptions(syncModuleOptions);
+    CompletableFuture<Void> stagedSyncersToRunSequentially =
+        CompletableFuture.completedFuture(null);
+
+    for (List<SyncModuleOption> syncOptions : groupedSyncModuleOptions) {
+      stagedSyncersToRunSequentially =
+          stagedSyncersToRunSequentially.thenCompose(
+              ignore ->
+                  chainSyncExecution(
+                      runnerNameOptionValue,
+                      isFullSync,
+                      isSyncProjectSyncCustomObjects,
+                      syncOptions));
+    }
+    return stagedSyncersToRunSequentially.whenComplete((syncResult, throwable) -> closeClients());
   }
 
   @Nonnull
-  CompletionStage<Void> sync(
-      @Nullable final String syncOptionValue,
+  private CompletableFuture<Void> chainSyncExecution(
       @Nullable final String runnerNameOptionValue,
       final boolean isFullSync,
-      final boolean isSyncProjectSyncCustomObjects) {
+      final boolean isSyncProjectSyncCustomObjects,
+      final List<SyncModuleOption> syncOptions) {
+    final List<CompletableFuture<Void>> syncersToRunParallel = new ArrayList<>();
+    for (SyncModuleOption syncOptionValue : syncOptions) {
+      Syncer<
+              ? extends ResourceView,
+              ?,
+              ? extends BaseSyncStatistics,
+              ? extends BaseSyncOptions<?, ?>,
+              ? extends QueryDsl<?, ?>,
+              ? extends BaseSync<?, ?, ?>>
+          syncer =
+              buildSyncer(syncOptionValue, runnerNameOptionValue, isSyncProjectSyncCustomObjects);
+      syncersToRunParallel.add(
+          syncer.sync(runnerNameOptionValue, isFullSync).toCompletableFuture());
+    }
+    return CompletableFuture.allOf(syncersToRunParallel.toArray(new CompletableFuture[0]));
+  }
+
+  @Nonnull
+  private static List<SyncModuleOption> validateAndCollectSyncOptionValues(
+      @Nonnull final String[] syncOptionValues) {
 
-    if (isBlank(syncOptionValue)) {
+    if (null == syncOptionValues) {
```

Comment:
with `Nonnull` annotation, it can not be null so the check looks a bit weird, either param should be nullable or the check is not necessary?

Reply:
This check was unnecessary. Removed and added a blank value validation inside the loop. https://github.com/commercetools/commercetools-project-sync/commit/94e2235952eb3940c4339283946bca03b8d20475